nixos/modules/services/networking/microsocks.nix

```
@@ -0,0 +1,140 @@
+    port = mkOption {
+      type = types.port;
+      default = 1080;
+      description = lib.mdDoc "Port on which microsocks should listen.";
+    };
+    disableLogging = mkOption {
+      type = types.bool;
+      default = false;
+      description = lib.mdDoc "If true, microsocks will not log any messages to stdout/stderr.";
+    };
+    authOnce = mkOption {
+      type = types.bool;
+      default = false;
+      description = lib.mdDoc ''
+        If true, once a specific ip address authed successfully with user/pass,
+        it is added to a whitelist and may use the proxy without auth.
+      '';
+    };
+    outgoingBindIp = mkOption {
+      type = types.nullOr types.str;
+      default = null;
+      description = lib.mdDoc "Specifies which ip outgoing connections are bound to";
+    };
+    authUsername = mkOption {
+      type = types.nullOr types.str;
+      default = null;
+      example = "alice";
+      description = lib.mdDoc "Optional username to use for authentication.";
+    };
+    authPasswordFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      example = "/run/secrets/microsocks-password";
+      description = lib.mdDoc "Path to a file containing the password for authentication.";
+    };
+    execWrapper = mkOption {
+      type = types.nullOr types.str;
+      default = null;
+      example = ''
+        ''${pkgs.mullvad-vpn}/bin/mullvad-exclude
+      '';
+      description = lib.mdDoc ''
+        An optional command to prepend to the microsocks command (such as proxychains, or a VPN exclude command).
+      '';
+    };
+  };
+  config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = (cfg.authUsername != null) == (cfg.authPasswordFile != null);
+        message = "Need to set both authUsername and authPasswordFile for microsocks";
+      }
+    ];
+    users = {
+      users = mkIf (cfg.user == "microsocks") {
+        microsocks = {
+          group = cfg.group;
+          isSystemUser = true;
+        };
+      };
+      groups = mkIf (cfg.group == "microsocks") {
+        microsocks = {};
+      };
+    };
+    systemd.services.microsocks = {
+      enable = true;
+      description = "a tiny socks server";
+      after = ["network.target"];
+      wantedBy = ["multi-user.target"];
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+        Restart = "on-failure";
+        RestartSec = 10;
+        LoadCredential = lib.optionalString (cfg.authPasswordFile != null) "MICROSOCKS_PASSWORD_FILE:${cfg.authPasswordFile}";
+        MemoryDenyWriteExecute = true;
+        SystemCallArchitectures = "native";
+        PrivateTmp = true;
+        NoNewPrivileges = true;
+        ProtectSystem = "strict";
+        ProtectHome = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        PrivateDevices = true;
+        RestrictSUIDSGID = true;
+        RestrictNamespaces = ["uts" "ipc" "pid" "user" "cgroup"];
```

```suggestion
        RestrictNamespaces = [ "uts" "ipc" "pid" "user" "cgroup" ];
```
As always, add spaces in `[` and `]`.
Also whenever possible, default to alphabetically ordering items in lists and attrs.
Exception is, `meta.broken` is always last because to avoid git conflicts. (It's a field that changes more often.)
I wish we had a linter, a formatter, that would get things right. No kidding. It's very debatable. And I having to give back such feedback. I'm just parroting what I was told when I first got here.

---

A formatter would be nice. All the nits you've identified are table stakes for any formatter.

It's quite a shame that [Alejandra](https://github.com/kamadorueda/alejandra)
isn't formatting how you'd like since it seems to be the most common formatter. 

Anyways I'll fix them up.

---

Let's settle for Alejandra.

---

There is also: https://hackage.haskell.org/package/nixfmt

https://github.com/NixOS/rfcs/pull/166